FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc/arc_builtin.h

```diff
@@ -0,0 +1,154 @@
+/* ------------------------------------------
+ * Copyright (c) 2022, Synopsys, Inc. All rights reserved.
+
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+
+ * 1) Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+
+ * 2) Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+
+ * 3) Neither the name of the Synopsys, Inc., nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+   --------------------------------------------- */
+
+/**
+ * @file
+ * @ingroup ARC_HAL_BUILTIN
+ * @brief Header file of builtin and helper functions
+ *
+ * The Metaware toolchain and the GNU toolchain are supported. The details please go to see the file.
+ */
+
+/**
+ * @addtogroup	ARC_HAL_BUILTIN
+ * @{
+ */
+
+#ifndef _ARC_HAL_BUILTIN_H_
+#define _ARC_HAL_BUILTIN_H_
+
+#include "arc.h"
+
+#ifndef __ASSEMBLY__
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @fn uint32_t arc_aux_read(uint32_t aux)
+ * @brief Read auxiliary register
+ *
+ * @param aux auxiliary register address
+ * @return value of auxiliary register
+ */
+Inline uint32_t arc_aux_read(uint32_t aux)
+{
+	uint32_t ret;
+
+	Asm("lr %0, [%1]"
+		: "=r" (ret)
+		: "r" (aux));
+
+	return ret;
+}
+
+/**
+ * @fn void arc_aux_write(uint32_t aux, uint32_t val)
+ * @brief Write auxiliary register
+ *
+ * @param aux auxiliary register address
+ * @param val Value to write
+ */
+Inline void arc_aux_write(uint32_t aux, uint32_t val)
+{
+	Asm(
+		"sr   %0, [%1]	\n"
+		:
+		: "ir" (val), "r" (aux));
+}
```

Comment:
We can use builtins here instead of ASM:

```
#if defined(__CCAC__)

#define z_arc_v2_aux_reg_read(reg) _lr((volatile uint32_t)reg)
#define z_arc_v2_aux_reg_write(reg, val) _sr((unsigned int)val, (volatile uint32_t)reg)

#else /* ! __CCAC__ */

#define z_arc_v2_aux_reg_read(reg) __builtin_arc_lr((volatile uint32_t)reg)
#define z_arc_v2_aux_reg_write(reg, val) __builtin_arc_sr((unsigned int)val, (volatile uint32_t)reg)

#endif /* __CCAC__ */
```